src/pip/_internal/resolution/resolvelib/candidates.py

```diff
@@ -120,7 +132,34 @@ def dist(self):
 
     def get_dependencies(self):
         # type: () -> Sequence[InstallRequirement]
-        return [
+        dependencies = [
             self._make_install_req(str(r), self._ireq)
             for r in self.dist.requires()
         ]
+        requires_python = get_requires_python(self.dist)
+        if requires_python is not None:
+            python_req = self._make_python_req(requires_python)
+            raise NotImplementedError(
+                "TODO: How do we insert this req?", python_req,
```

Comment:
My instinct is that we should return `Requirement` objects rather than `InstallRequirement` objects from this method. I've made that assumption by mistake a couple of times now.

I think this should be possible by passing around partially-applied functions rather than objects that just get passed as arguments into other functions, but I'll need to take a bit of time to get my head around the data flows just to be sure of that.

Reply:
This sort of ties back to the `make_candidate` growing comment as well. Maybe we need to have something to hold all these partial functions to simplify the signatures. A… provider? (boo, I’m running out of names)

Reply:
Resolved with the factory refactor, now this returns `Sequence[Requirement]`.